internal/extproc/bodymutator/body_mutator.go

```diff
@@ -0,0 +1,123 @@
+// Copyright Envoy AI Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package bodymutator
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/tidwall/sjson"
+
+	"github.com/envoyproxy/ai-gateway/internal/filterapi"
+)
+
+type BodyMutator struct {
+	// originalBody is the original request body for retry scenarios
+	originalBody []byte
+
+	// bodyMutations is the body mutations to apply
+	bodyMutations *filterapi.HTTPBodyMutation
+}
+
+func NewBodyMutator(bodyMutations *filterapi.HTTPBodyMutation, originalBody []byte) *BodyMutator {
+	return &BodyMutator{
+		originalBody:  originalBody,
+		bodyMutations: bodyMutations,
+	}
+}
+
+// isJSONValue checks if a string represents a JSON value (not a plain string)
+func isJSONValue(value string) bool {
+	value = strings.TrimSpace(value)
+
+	// Check for quoted strings (JSON strings)
+	if strings.HasPrefix(value, "\"") && strings.HasSuffix(value, "\"") {
+		return true
+	}
+
+	// Check for numbers (integers or floats)
+	if value == "0" || value == "true" || value == "false" || value == "null" {
+		return true
+	}
+
+	// Check for positive/negative numbers
+	if len(value) > 0 {
+		first := value[0]
+		if (first >= '0' && first <= '9') || first == '-' || first == '+' {
+			// Simple number check - contains only digits, dots, +, -, e, E
+			isNumber := true
+			for _, r := range value {
+				if (r < '0' || r > '9') && r != '.' && r != '-' && r != '+' && r != 'e' && r != 'E' {
+					isNumber = false
+					break
+				}
+			}
+			if isNumber {
+				return true
+			}
+		}
+	}
+
+	// Check for objects or arrays
+	if strings.HasPrefix(value, "{") && strings.HasSuffix(value, "}") {
+		return true
+	}
+	if strings.HasPrefix(value, "[") && strings.HasSuffix(value, "]") {
+		return true
+	}
+
+	// Default to plain string
+	return false
+}
+
+// Mutate mutates the request body based on the body mutations and restores original body if mutated previously.
+func (b *BodyMutator) Mutate(requestBody []byte, onRetry bool) ([]byte, error) {
+	if b.bodyMutations == nil {
+		return requestBody, nil
+	}
+
+	if onRetry && b.originalBody != nil {
+		// On retry, restore the original body first
+		requestBody = b.originalBody
+	}
+
+	mutatedBody := requestBody
+	var err error
+
+	// Apply removals first
+	if len(b.bodyMutations.Remove) > 0 {
+		for _, fieldName := range b.bodyMutations.Remove {
+			if fieldName != "" {
+				mutatedBody, err = sjson.DeleteBytes(mutatedBody, fieldName)
```

Comment:
so DeleteBytes by default copies the the underlying buffer, so let's put a TODO comment here too saying we should optimize by using SetBytesOptions with nil value + replaceInPlace tracking like below